File: 2021/R/d02.R
parse_input <- function(d) {
  read.csv(d, header = FALSE, sep = " ", col.names = c("dir", "val"))
}

summer <- function(d,x) {
  sum(d$val[d$dir == x])
}

part1 <- function(d) {
  summer(d, "forward") * (summer(d, "down") - summer(d, "up"))
}

part2 <- function(d) {
  dep <- 0
  aim <- 0
  for (i in seq_len(nrow(d))) {
    if (d$dir[i] == "forward") {
      dep <- dep + (aim * d$val[i])
    } else {
      aim <- aim + (d$val[i] * ifelse(d$dir[i] == "down", 1, -1))
    }
  }
  summer(d, "forward") * dep
}

d <- parse_input("../inputs/input_2.txt")
c(part1(d), part2(d))

File: 2021/PHP/d02.php
<?php

  function solve($d, $p) {
    $d = explode("\n", $d);
    $hrz = 0;
    $dep = 0;
    $aim = 0;
    for ($i = 0; $i < count($d); $i++) {
      $cmd = explode(" ", $d[$i]);
      if (count($cmd) < 2) continue;
      $val = intval($cmd[1]);
      $cmd = $cmd[0];
      if ($cmd == "forward") $hrz = $hrz + $val;
      if ($p == 1) {
        if ($cmd == "up") $dep = $dep - $val;
        else if ($cmd == "down") $dep = $dep + $val;
      } else {
        if ($cmd == "forward") $dep = $dep + ($aim * $val);
        else if ($cmd == "up") $aim = $aim - $val;
        else $aim = $aim + $val;
      }
    }
    return $dep * $hrz;
  }

  $d = file_get_contents("../inputs/input_2.txt");
  echo ("Part 1: ".solve($d, 1)."   Part 2: ".solve($d, 2)."\n");

?>